### PR TITLE
Orca: Add merge() function to SparkXShards and an NCF pytorch example with new features

### PR DESCRIPTION
**Pull request by zpeng1898**
Summary: 
To merge two SparkXShards like `pandas.DataFrame.merge()`, add the function `merge()` to SparkXShards in the file `bigdl.orca.data.shard.py`. Also, a NCF pytorch example `train_xshards_add_features.py` is added to demonstrate how to use `merge()` API to generate new features by merging two XShards.

Details: 
1.`merge()` API
The [issue ](https://github.com/analytics-zoo/orca/issues/102#issuecomment-1227185777)fully describes the purpose of the API.
The style of this API is similar to [pandas.DataFrame.merge()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html), and it calls [spark.DataFrame.join()](https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.DataFrame.join.html).

2.`train_xshards_add_features.py`
The example uses two `SparkXShards` as model input and calls `SparkXShards.merge()` to combine their features together.

### CODE REVIEW

File: python/orca/src/bigdl/orca/data/shard.py [hunk 1]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
 
         return self.rdd.map(lambda x: func(x)).first()
 
+    def merge(self,right, how="inner", on=None,**kwargs):
+        """
+        Merge two SparkXShards based on pyspark.sql.DataFrame.join()
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:709]: Merge two SparkXShards into a single SparkXShards with a database-style join.

File: python/orca/src/bigdl/orca/data/shard.py [hunk 2]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
...
+        
+        :param right: The other SparkXShards to be merged with self
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:711]: remove `with self`

File: python/orca/src/bigdl/orca/data/shard.py [hunk 3]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
...
+        :param how: 'left','right','outer' or 'inner'. Type of merge. Default is 'inner'.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:712]: Type of merge, 'left','right','outer' or 'inner'. Default is 'inner'.

File: python/orca/test/bigdl/orca/data/test_spark_backend.py [hunk 4]
```diff
@@ -251,6 +251,24 @@ def test_standardscale_shards(self):
...
+            ])
+        df2=spark.createDataFrame([
+            Row(a=1, c=7),
+            Row(a=2, c=8),
+            Row(a=4, c=9)
+            ])
+        data_shard1 = spark_df_to_pd_sparkxshards(df1)
+        data_shard2 = spark_df_to_pd_sparkxshards(df2)
+        merged_shard = data_shard1.merge(data_shard2, on='a')
+        assert len(merged_shard)==2
```
Comment (hkvision) [inline python/orca/test/bigdl/orca/data/test_spark_backend.py:270]: Need to test all the results

File: python/orca/src/bigdl/orca/data/shard.py [hunk 5]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
...
+        :param on: Column or index level names to merge on
+        :return: A new merged SparkXShards
+        """
+        from bigdl.orca.data.utils import spark_df_to_pd_sparkxshards
+        invalidInputError(isinstance(right, SparkXShards), "right should be a SparkXShards")
+        
+        left_df = self.to_spark_df()
+        right_df = right.to_spark_df()
+        merged = left_df.join(right_df,on=on,how=how)
+        merged = merged.limit(merged.count())
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:722]: why this?

File: python/orca/src/bigdl/orca/data/shard.py [hunk 6]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
 
         return self.rdd.map(lambda x: func(x)).first()
 
+    def merge(self,right, how="inner", on=None,**kwargs):
+        """
+        Merge two SparkXShards based on pyspark.sql.DataFrame.join()
+        
+        :param right: The other SparkXShards to be merged with self
+        :param how: 'left','right','outer' or 'inner'. Type of merge. Default is 'inner'.
+        :param on: Column or index level names to merge on
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:713]: Column name(s) to join on.

File: python/orca/src/bigdl/orca/data/shard.py [hunk 7]
```diff
@@ -704,6 +704,26 @@ def func(pdf):
...
+        """
+        Merge two SparkXShards into a single SparkXShards with a database-style join.
+        
+        :param right: The other SparkXShards to be merged.
+        :param how: Type of merge. 'left', 'right', 'outer' or 'inner'. Default is 'inner'.
+        :param on: Column name(s) to join on.
+        :return: A new merged SparkXShards.
+        """
+        from pyspark.sql import SparkSession
+        spark = SparkSession.builder.getOrCreate()
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:717]: do you still need this?

File: python/orca/tutorial/pytorch/NCF/train_xshards_add_features.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
+#
```
Comment (hkvision) [inline python/orca/tutorial/pytorch/NCF/train_xshards_add_features.py:1]: remove this file from this PR

File: python/orca/src/bigdl/orca/data/shard.py (hunk 7 above)
Comment (zpeng1898) [reply python/orca/src/bigdl/orca/data/shard.py]: The line has been deleted.

**Review by hkvision — APPROVED** [review]

File: python/orca/src/bigdl/orca/data/shard.py [hunk 9]
```diff
@@ -704,6 +704,25 @@ def func(pdf):
...
+        :param on: Column name(s) to join on.
+        :return: A new merged SparkXShards.
+        """   
+        from bigdl.orca.data.utils import spark_df_to_pd_sparkxshards
+        invalidInputError(isinstance(right, SparkXShards), "right should be a SparkXShards")
+
+        left_df = self.to_spark_df()
+        right_df = right.to_spark_df()
+        merged = left_df.join(right_df, on=on, how=how)
+        merged = merged.coalesce(min(merged.count(), merged.rdd.getNumPartitions()))
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/data/shard.py:722]: @dding3 is this correct?

Comment (dding3) [reply python/orca/src/bigdl/orca/data/shard.py:722]: Please change it to `merged.coalesce(left_df.rdd.getNumPartitions())`

**Comment by zpeng1898** [comment]
image
The jenkins #1166 has passed.

**Comment by zpeng1898** [comment]
To merge two SparkXShards like pandas.DataFrame.merge(), add the function merge() to SparkXShards in the file bigdl.orca.data.shard.py.

**Review by dding3 — APPROVED** [review]

File: python/orca/src/bigdl/orca/data/shard.py [hunk 10]
```diff
@@ -816,6 +816,41 @@ def func(pdf):
...
+        nonEmptyPart = spark.sparkContext.accumulator(0)
+        def f(iterator):
+            isEmpty = 1
+            for x in iterator:
+                if(isEmpty == 0):
+                    break
+                else:
+                    isEmpty = 0
+            nonEmptyPart.add(isEmpty == 0)
+        merged.rdd.foreachPartition(f)
```
Comment (dding3) [inline python/orca/src/bigdl/orca/data/shard.py:846]: One optimization we may add here is `if there is no empty partition, then we will skip the repartition operation`

**Review by hkvision — APPROVED** [review]

**Comment by zpeng1898** [comment]
http://10.112.231.51:18888/job/ZOO-PR-BigDL-Python-Spark-3.1-py37-ray/1282/console

**Pull request merged**